File: Data Structures/Python_JS - Data Structures/Graphs/QJs/graph.js
log = (val) => console.log(val)

class Graph {
    constructor() {
        this.adjacencyList = {}
    }

    addVertex(name) {
        if(!this.adjacencyList[name]) this.adjacencyList[name] = []
    }

    addEdge(vertex1, vertex2) {
        !!this.adjacencyList[vertex1] && !!this.adjacencyList[vertex2] && this.adjacencyList[vertex1].push(vertex2)
        !!this.adjacencyList[vertex2] && !!this.adjacencyList[vertex1] && this.adjacencyList[vertex2].push(vertex1)        
    }

    removeEdge(vertex1, vertex2) {
        !!this.adjacencyList[vertex1] &&
        (this.adjacencyList[vertex1] = this.adjacencyList[vertex1].filter(
            c => c != vertex2
        ))

        !!this.adjacencyList[vertex2] &&
        (this.adjacencyList[vertex2] = this.adjacencyList[vertex2].filter(
            c => c != vertex1
        ))
    }

    removeVertex(vertex) {
        while(this.adjacencyList[vertex].length) {
            const removed = this.adjacencyList[vertex].pop()
            this.removeEdge(removed, vertex)
        }
        delete this.adjacencyList[vertex]
    }

    dfsRecursive(vertex, results = [], visited = {}) {
        if(!this.adjacencyList[vertex]) return
        results.push(vertex)
        visited[vertex] = true
        for(let n of this.adjacencyList[vertex]) {
            if(!visited[n]) {
                this.dfsRecursive(n, results, visited)
            }
        }

        return results
    }

    dfsIterative(vertex) {
        let s = []
        let visited = []
        let result = []
        s.push(vertex)
        while(!!s.length) {
            const node = s.pop()
            if(!visited[node]) {
                result.push(node)
                visited[node] = true
                for(let n of this.adjacencyList[node]) {
                    s.push(n)
                }
            }
        }

        return result
    }

    dfs(start) {
        let queue = [start]
        let results = []
        let visited = {}

        visited[start] = true
        while(!!queue.length) {
            let first = queue.shift()
            results.push(first)
            for(let n of this.adjacencyList[first]) {
                if(!visited[n]) {
                    visited[n] = true
                    queue.push(n)
                }
            }
        }

        return results
    }
}

let g = new Graph()
g.addVertex('A')
g.addVertex('B')
g.addVertex('C')
g.addVertex('D')
g.addVertex('E')
g.addVertex('F')

g.addEdge('A', 'B')
g.addEdge('A', 'C')
g.addEdge('B', 'D')
g.addEdge('C', 'E')
g.addEdge('D', 'E')
g.addEdge('D', 'F')
g.addEdge('E', 'F')

log(g.dfsRecursive('A'))
//log(g.dfsIterative('A'))

File: Practice Data Structures/Python_JS - Data Structures/Graphs/AJs/graph.py
# addVertex, addEdge, removeVertex, removeEdge
class Graph:
    def __init__(self):
        self.adjacencyList = {}
    
    def __str__(self):
        return str(self.__class__) + ": " + str(self.__dict__)
    
    def addVertex(self, name):
        # adds an unconnected node
        if (name in self.adjacencyList) is False: self.adjacencyList[name] = []
    
    def addEdge(self, vertex1, vertex2):
        if ((vertex1 in self.adjacencyList) is False) | ((vertex2 in self.adjacencyList) is False): return False
        self.adjacencyList[vertex1].append(vertex2)
        self.adjacencyList[vertex2].append(vertex1)
    
    def removeVertex(self, vertex):
        # to remove a vertex you pop out every connection it has
        # and then remove the edge between them
        node = self.adjacencyList[vertex]
        if node is None: return None
        while len(node) > 0:
            removed = node.pop()
            self.removeEdge(removed, vertex)
        self.adjacencyList.pop(vertex)
    
    def removeEdge(self, vertex1, vertex2):
        if (vertex1 in self.adjacencyList):
            self.adjacencyList[vertex1] = list(filter(lambda item: item != vertex2, self.adjacencyList[vertex1]))
        if (vertex2 in self.adjacencyList):
            self.adjacencyList[vertex2] = list(filter(lambda item: item != vertex1, self.adjacencyList[vertex2]))

    def dfsIterative(self, vertex):
        # we're picking a node
        # if we haven't visited that node before
        # we add it as visited
        # we pick all its connections, and repeat the above operation
        state = []
        result = []
        visited = {}
        state.insert(len(state), vertex)
        while len(state) > 0:
            node = state.pop()
            if visited.__contains__(node) is False:
                visited[node] = True
                result.insert(len(result), node)
                for vertx in self.adjacencyList[node]:
                    state.insert(len(state), vertx)
            
        return result

    def dfsRecursive(self, vertex, result=[], visited={}):
        if (vertex in self.adjacencyList) is False: return None
        result.insert(len(result), vertex)
        visited[vertex] = True
        for vertx in self.adjacencyList[vertex]:
            if (vertx in visited) is False:
                self.dfsRecursive(vertx, result, visited)
        return result

g = Graph()
g.addVertex('A')
g.addVertex('B')
g.addVertex('C')
g.addVertex('D')
g.addVertex('E')
g.addVertex('F')

g.addEdge('A', 'B')
g.addEdge('A', 'C')
g.addEdge('B', 'D')
g.addEdge('C', 'E')
g.addEdge('D', 'E')
g.addEdge('D', 'F')
g.addEdge('E', 'F')

print(g.adjacencyList)
print()
print(g.dfsRecursive('A'))
print(g.dfsIterative('A'))


            
